Color spaces C1C2C3 and MAXRGB added.

imransalam opened the pull request:
## Description
This is an addition to color spaces, I have added two color spaces, i-e C1C2C3 and MAXRGB. 

![figure_1-1](https://user-images.githubusercontent.com/12950439/33764634-17219f80-dc37-11e7-8636-c46f9dba5622.png)

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [ Yes ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ Yes ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ Could Not find any examples of color conversions in ./doc/examples] Gallery example in `./doc/examples` (new features only)
- [ Yes ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]


## References
https://arxiv.org/abs/1702.05421
https://docs.gimp.org/en/plug-in-max-rgb.html

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

imransalam committed:
rgb2c1c2c3 and rgb2maxrgb added in color convertions

imransalam committed:
__init__.py updated

imransalam committed:
Smoothing Factor added in rgb2c1c2c3

imransalam committed:
Deep copying issue in max rgb resolved

pep8speaks commented:
Hello @imransalam! Thanks for submitting the PR.



 - In the file [`skimage/color/colorconv.py`](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py), following are the PEP8 issues :

> [Line 689:1](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L689): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 718:24](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 718:54](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 718:66](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 718:70](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 718:80](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (180 > 79 characters)
> [Line 718:102](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 718:114](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 718:118](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 718:150](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 718:162](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 718:166](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L718): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 724:1](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L724): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (4)
> [Line 754:14](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L754): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 754:16](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L754): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 755:14](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L755): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 755:16](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L755): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 756:14](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L756): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 756:16](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L756): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 761:23](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L761): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 761:25](https://github.com/scikit-image/scikit-image/blob/579bca8d0d6edd2e6d436305e18427d324d32d3d/skimage/color/colorconv.py#L761): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

imransalam committed:
Changes made wrt to PEP8 Standards

imransalam commented:
@pep8speaks Hi, I've updated the skimage/color/colorconv.py with PEP8 standards in the latest commit.

hmaarrfk commented:
This looks really cool.

Sorry that this slipped the radar. Are you still interested in adding this? It would definitely need some tests that ensure:
1. The mathematical correctness.
2. The range of inputs (Grayscale vs RGB)  we care for are adequately addressed.

imransalam commented:
@hmaarrfk Hey. Sure let me know what tests you want to perform.

hmaarrfk commented:
Cool, take a look at 
https://github.com/scikit-image/scikit-image/blob/master/skimage/color/tests/test_colorconv.py#L497

essentially, you will want to add your function, each prefixed with `test`.

Probably in that same directory.

Don't worry about putting them in a `class` that is some old stuff that was required by `nose`.

Generally, you might want to find an other way to compute the result. That could be:
1. finding a less efficient way of computing it.
2. Copy pasting the output of a manually checked result.

See what you can think of on the test side.

hmaarrfk commented on skimage/color/colorconv.py:
```diff
@@ -41,7 +41,7 @@
 :author: Travis Oliphant (XYZ and RGB CIE functions)
 :author: Matt Terry (lab2lch)
 :author: Alex Izvorski (yuv2rgb, rgb2yuv and related)
-
```
I think generally we are now adding each person's contributions to the file `CONTRIBUTORS.txt`.

@stefanv should Imram add his name here too for this module?

hmaarrfk commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+
+    References
+    ----------
+    .. [1] https://arxiv.org/abs/1702.05421
```
This reference needs to be fleshed out some more. I wish we could convert bib files to our format easily.

hmaarrfk commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+    --------
+    >>> from skimage import data
+    >>> img = data.astronaut()
+    >>> img_c1c2c3 = rgb2c1c2c3(img)
```
This example is probably better shown in a gallery.

hmaarrfk commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+
+    References
+    ----------
+    .. [1] https://docs.gimp.org/en/plug-in-max-rgb.html
```
This should at least be given a title and a retrieved on time stamp.

hmaarrfk commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+
+    Examples
+    --------
+    >>> from skimage import data
```
See if you can adapt your code to fit in a new file in this directory.

https://github.com/scikit-image/scikit-image/tree/master/doc/examples/color_exposure

hmaarrfk commented:
@imransalam before working on anything, I would see if anybody else from @scikit-image/core has objection to adding this kind of functionality. Color space has been quite contentious here as the API has been inconsistent and some have even proposed moving to an other library. I cant see to recall the exact conversation though.

Borda commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
 
+
+def rgb2c1c2c3(rgb):
+    """RGB to C1C2C3 color space conversion.
```
I would welcome a short decribtion of what the C1C2C3 is... giving reference is nice but keep it easy to use :)

Borda commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+    >>> img_c1c2c3 = rgb2c1c2c3(img)
+    """
+    out = np.arctan(rgb / np.dstack((
+        np.maximum(rgb[..., 1], rgb[..., 2]) + 0.000000001, 
```
rather use `np.finfo(float).eps`
https://het.as.utexas.edu/HET/Software/Numpy/reference/generated/numpy.finfo.html

Borda commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+
+
+def rgb2maxrgb(rgb):
+    """RGB to MAXRGB color space conversion.
```
I would welcome a short description of what the MAXRGB is... giving reference is nice but keep it easy to use :)

Borda commented on skimage/color/colorconv.py:
```diff
@@ -679,6 +679,85 @@ def rgb2xyz(rgb):
+    R = rgb[:, :, 0]
+    G = rgb[:, :, 1]
+    B = rgb[:, :, 2]
+    M = np.maximum( np.maximum(R, G), B)
```
`np.max(rgb, axis=-1)` is easier to read/undestand

imransalam committed:
Updated Description, suggested code changes

Updated the description for C1C2C3 and MaxRGB. Added np.finfo(float).eps instead of adding a small floating point myself. Used the max with axis=-1.

imransalam committed:
Added test cases for maxrgb and c1c2c3

Added test cases for both maxrgb and c1c2c3. Manual testing.

imransalam committed:
resolving pep8 issues

imransalam committed:
solved pep8 issues

imransalam committed:
solved pep8 issues

imransalam committed:
Merge branch 'master' into master

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
